router/module.go

```diff
@@ -278,10 +279,27 @@ func (am AppModule) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, re
 
 // OnAcknowledgementPacket implements the IBCModule interface
 func (am AppModule) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, acknowledgement []byte, relayer sdk.AccAddress) error {
-	return am.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	if err := am.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer); err != nil {
+		return err
+	}
+
+	var ackErr channeltypes.Acknowledgement_Error
+	if err := json.Unmarshal(acknowledgement, &ackErr); err == nil && len(ackErr.Error) > 0 {
+		am.keeper.RefundForwardedPacket(ctx, packet)
+	}
+
+	return nil
```

Comment:
If we receive an ack without an error. Shouldn't we remove the `inFlightPacket` from storage?

Reply:
good catch, updated